content/docs/tutorials/handling-errors.mdx

```diff
@@ -0,0 +1,281 @@
+
+
+### Timeouts
+If you receive a 504 error from Horizon after a transaction submission, you've encountered a timeout. In the Stellar universe, a timeout indicates something a little nebulous and non-traditional: it's more of a *warning* that your request hasn't been fulfilled within a reasonable amount of time *yet* more than an error. This subtlety arises because of the nature of the relationship between Horizon and Stellar Core: the network might take some time to accept a transaction---especially one with a low fee, see [Surge Pricing](#surge-pricing) later---on the order of 5-10 minutes in the worst cases, whereas Horizon needs to provide developers with a response within 30 seconds or so.
+
+This leads to a very important point: **receiving a 504 for your transaction submission does not mean the transaction did not make it to the network**.
+
+There's no mechanism to "cancel" a transaction after it has been submitted aside from by exceeding its timebounds. In fact, one of the best tools available for managing transaction failures is through [timebounds](https://developers.stellar.org/docs/glossary/transactions/#time-bounds). While optional, they're *crucial* in simplifying error handling: it's much easier to know when to check a transaction's status when you know the maximum amount of time it can be stuck.
+
+With this trick in our back pocket, resolving 504s is straightforward: until the timebound has expired, you can safely retry submission *without modifying the transaction*. Horizon makes an important guarantee (excerpt is from [this example](https://developers.stellar.org/docs/tutorials/send-and-receive-payments/#send-a-payment)) for this scenario:
+
+> If the transaction has already been successfully applied to the ledger, Horizon will simply return the saved result and not attempt to submit the transaction again. Only in cases where a transaction’s status is unknown (and thus will have a chance of being included into a ledger) will a resubmission to the network occur.
+
+In other words, if your transaction made it through after the 504, you'll get your previously-submitted transaction back. If it didn't (e.g. a `tx_too_late` result code), you're essentially back to square one: you might 504 again, fail for another reason (e.g. `tx_bad_seq`), or succeed.
+
+#### Transaction Bans
+Note the following caveat of the [transaction lifecycle](https://developers.stellar.org/docs/glossary/transactions/#transaction-lifecycle): 
+
+> If a transaction is kept in memory after a certain number of ledger closes, **it will be banned for several additional ledgers.** This means no attempt will be made to include it in a candidate transaction set additional ledgers during this time.
```

Comment:
How does one end up causing this situation? If I get a `504`, and I keep resubmitting, am I risking the transaction getting banned?

Reply:
From my understanding, bans happen if a tx stays in the queue for too long (others are taking priority due to higher fees). I've clarified this!

Reply:
So as I understand, a transaction will stay in the queue until one of the following happens:
- it is included in the ledger (success)
- its upper timebound is exceeded (tx_too_late)
- its banned

Please correct me if I'm wrong but this raises the following questions:
- is there a fixed limit on how long a tx can stay in the queue (# of ledgers / # of seconds)? This limit should be mentioned if it exists.
- what is the Horizon response when I try to submit a banned transaction? I don't see a result code that directly applies
- Lets say every transaction submitted has a higher fee than the one I submit, and I don't include an upper timebound. The TX _will_ be banned right?

Reply:
Oh I see you answered my first question here: https://github.com/stellar/new-docs/pull/328#discussion_r579516471

But you say "typically". If I want to avoid getting banned (because I don't want to pay the 10x required to be reconsidered), what guidance would you give me? Always set an upper timebound of no more than 15 seconds (~3 ledgers)? I think we should provide some guidance here in the doc, because getting banned is a significant penalty.

Reply:
I believe there's no limit on staying in the queue aside from what's achievable via the tx timebound. Here's what @MonsieurNicolas had to say on the matter:

> You need to set your timeouts based on your business requirements and assume that transactions may get dropped or re-processed at any time (transactions may sit around in queues for much longer than 3 ledgers). The only thing you can rely on are close time in the latest ledger that was closed and sequence number of your account in that latest ledger, then you need to design a sequence of transactions that deals with those races properly.

If you want to avoid a ban, the best strategy is (_likely_-I'm doing some inferring here) just to set your fees appropriately. For example, I can't imagine setting a max fee based on the 90th percentile (from `/fee_stats`) will ever lead to getting banned. I'll try to get answers for your other bullets from Core.

Reply:
Despite our mutual curiosity and desire for clarity, @JakeUrban, I think it's actually best to avoid any mention of tx bans in this guide entirely (see fd87d6b68cb7f4ccd6748c58f57c05dd6d94e20d).

This thought process is magnified by the fact that a txban-specific error doesn't bubble up to SDK users and so represents an error scenario that's fairly deep and specific. As far as users are concerned, the resolution strategy stays the same: **set the max fee you're willing to pay** (or make a fee-bump / check `/fee_stats`) **and try again**. I think "exposing" this (deep) knowledge about Stellar Core's behavior (a) isn't actually helpful (except to satisfy curiosity) (b) strips away more of the (intended) abstraction of Horizon <--> Core and (c) locks Core into certain expectations. 

What do you think?

Reply:
I don't think we need to delay merging this doc over this conversation, but I do think it exposed some gaps in our understanding. 

It's not clear how Horizon handles Core's `TRY_AGAIN_LATER` error and how that is relayed to the client. It may be the case that Horizon lumps in into other error modes described in this doc (`504`), in which case the client truly doesn't need to worry about it. But I wanted to confirm that there isn't some undocumented error response that represents a tx ban that clients wouldn't be prepared to handle.